### PR TITLE
feat(daisi): Use new scenariofile parser structure for minhton

### PR DESCRIPTION
**Pull request by ltoenning**
Related to https://github.com/iml130/sola/issues/84.

# Proposed Changes
Use new scenariofile parser structure for MINHTON

The documentation for the scenariofile will be adjusted on the ``docs/daisi`` branch. 

# Definition of Done (optional)
- [x] Feature is implemented
- [x] No known bugs that stops the correct execution
- [x] CI / CD pipeline passed
- [x] The code guidelines were followed
- ~[ ] Unit and / or Integration tests for the new feature~
- [ ] New feature was added to the documentation

### CODE REVIEW

File: daisi/src/minhton-ns3/peer_discovery_attributes.h [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+  void parse(YAML::Node node);
+};
+
+/*******************/
+/* CONTENT BEHAVIOR */
+/*******************/
+
+struct ChoiceValue {
+  // Only one is/must valid
```
Comment (palask) [inline daisi/src/minhton-ns3/peer_discovery_attributes.h:42]:
```suggestion
  // Only exactly one must be valid
```

File: daisi/src/minhton-ns3/peer_discovery_attributes.h [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+  std::vector<ChoiceValue> values;
+
+  void parse(YAML::Node node);
+};
+
+struct Constant {
+  static std::string typeName() { return "constant"; }
+
+  // Only one is/must valid
```
Comment (palask) [inline daisi/src/minhton-ns3/peer_discovery_attributes.h:64]:
```suggestion
  // Only exactly one must be valid
```

File: daisi/src/minhton-ns3/scenario_steps.h [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+#define DAISI_MINHTON_NS3_MINHTON_SCENARIO_STEPS_H_
+
+#include <cstdint>
+#include <optional>
+#include <string>
+
+#include "manager/scenariofile_component.h"
+
+#define TYPE_NAME(name) \
+  static std::string typeName() { return name; }
```
Comment (palask) [inline daisi/src/minhton-ns3/scenario_steps.h:27]: Is there a particular reason why there is a macro `TYPE_NAME` here, but e.g. not in `daisi/src/minhton-ns3/peer_discovery_attributes.h`?

File: daisi/src/minhton-ns3/scenario_steps.h [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    SERIALIZE_VAR(delay);
+    SERIALIZE_VAR(level);
+    SERIALIZE_VAR(number);
+    SERIALIZE_VAR(index);
+  }
+};
+
+struct LeaveMany {
+  TYPE_NAME("leave-many");
+  std::optional<uint32_t> delay;
```
Comment (palask) [inline daisi/src/minhton-ns3/scenario_steps.h:80]: One free line in between please to keep it consistent and clean ;)

```suggestion
  TYPE_NAME("leave-many");
  
  std::optional<uint32_t> delay;
```

File: daisi/src/minhton-ns3/scenario_steps.h [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+  void parse(YAML::Node node) { SERIALIZE_VAR(delay); }
+};
+
+struct FindQuery {
+  TYPE_NAME("find-query");
+
+  uint64_t level;
+  uint64_t number;
```
Comment (palask) [inline daisi/src/minhton-ns3/scenario_steps.h:165]: Should probably be `uint32_t` just like in the other structs. And maybe think about whether validity_threshold actually needs to be a `uint64_t`.

File: daisi/utils/ci_scenario_files/minhton_cases.yml [hunk 6]
```diff
@@ -1,13 +1,15 @@
-# required
-title: Test
-name: Ein weiterer Test
-stoptime: 10000000
+title: MINHTON_default
```
Comment (palask) [inline daisi/utils/ci_scenario_files/minhton_cases.yml:1]:
```suggestion
title: MINHTON scenario including all join algorithm subcases
```

File: daisi/utils/ci_scenario_files/minhton_test.yml [hunk 7]
```diff
@@ -1,13 +1,15 @@
-# required
-title: Test
-name: Ein weiterer Test
-stoptime: 10000000
+title: MINHTON_default
```
Comment (palask) [inline daisi/utils/ci_scenario_files/minhton_test.yml:1]:
```suggestion
title: MINHTON default scenario
```

**Review by palask — APPROVED** [review]
Great job, nothing to complain about logic-wise, just some smaller adjustments may be useful.

File: daisi/src/minhton-ns3/scenario_steps.h (hunk 3 above)
Comment (ltoenning) [reply daisi/src/minhton-ns3/scenario_steps.h:27]: Not in particular. I now changed everything to directly implement ``typeName()`` and removed the macro.

File: daisi/src/minhton-ns3/scenario_steps.h (hunk 5 above)
Comment (ltoenning) [reply daisi/src/minhton-ns3/scenario_steps.h:165]: Think big :wink: 
No, I just changed this to ``uint32_t``. ``uint64_t`` was used by the old parser for all integer types, but 32-bit wide integers should be fine.

**Pull request merged**